### ISSUE TITLE
Safari Incognito - DOM Exception 22

### ISSUE THREAD

**Issue by Charizard** [Vestorly/torii#265]
Safari in incognito mode does not support `localStorage` so [this line](https://github.com/Vestorly/torii/blob/master/lib/torii/mixins/ui-service-mixin.js#L44) is throwing `DOM Exception 22`.

I believe there needs to be a fallback to another store?

**Comment by mixonic** [comment]
@Charizard this is correct, though we don't need to fall back to a store, we need to fall back to another cross-window communication channel. We had one before `localStorage` events that worked well. It should be fairly easy to look at that old code and wire up an abstraction that falls back.

/cc @bantic

**Comment by bantic** [comment]
Thanks @Charizard. This may have been an issue before, as well. It appears that localStorage has been in use since [at least 0.6.0](https://github.com/Vestorly/torii/blob/4547f9f1c61cb733ac119214bf4c9138295340c8/lib/torii/services/popup.js#L78).

Safari Private Browsing doesn't seem to offer any alternative that I can see for cross-window persistence, unfortunately. This nixes the ability to use popups with Torii in Safari Private Browsing. But the iframe service may be able to be used as a fallback if its usage of localStorage is changed.

**Comment by Charizard** [comment]
@bantic So the iframe service does not use any localStorage? You are saying it can be used in Safari Private browsing mode?

**Comment by mixonic** [comment]
@bantic I think one of our old strategies would work in safari private browsing.

Iframe and popups use the same communication channel.

**Comment by mixonic** [comment]
For example we used to use postMessage: https://github.com/Vestorly/torii/commit/5f7d19cf9f52aa1c736b1567bdb2c83f9506bd6a#diff-8b7b6e15c618f576b36e529bef3c661b

**Comment by mixonic** [comment]
The issue describing why we moved away from `postMessage`: https://github.com/Vestorly/torii/pull/195

**Comment by bantic** [comment]
> So the iframe service does not use any localStorage? You are saying it can be used in Safari Private browsing mode?

@Charizard it uses localStorage also; I was musing that it might be able to use a safe-for-private-browsing-safari option more easily, but neither one will work just yet. I've confirmed that postMessage works in Safari Private Browsing windows so it is a good candidate for a fallback for Torii to use.

**Comment by Charizard** [comment]
Looks like @mixonic has already found the [issue](https://github.com/Vestorly/torii/pull/195#issuecomment-115883438) :wink:

Since we decided if `localStorage` is not available torii would fallback to `postMessage`. If you can point me in the right direction, I might be able to give a PR. Sorry, I am new to the project.

**Comment by Charizard** [comment]
I have forked and applied a fallback to window messaging but against `0.6.1` which is what we are using in our project. If you can check this changes [here](https://github.com/Vestorly/torii/compare/4d33299c7156d7fe9a5d32097adf805c8a3df6a2...Charizard:falling-back-to-window-messaging) and give me your feedback, then I will be able to give a PR against master.

Thanks.

**Comment by anthonycollini** [comment]
Was there ever a PR opened for this? This thread is 8 months old.

**Comment by zbeat** [comment]
_bump_

**Comment by Charizard** [comment]
@anthonycollini No it was not opened coz the the fix I made for supporting Safari's Incognito mode was made on `0.6.1` (since it is what we are using) of Torii.

I can take some time and open up a PR by the end of this week.
